lib/binaries/config_source.ts

```diff
@@ -0,0 +1,189 @@
+import * as fs from 'fs';
+import * as os from 'os';
+import * as path from 'path';
+import * as request from 'request';
+import * as semver from 'semver';
+import * as xml2js from 'xml2js';
+
+import {Config} from '../config';
+import {HttpUtils} from '../http_utils';
+
+export abstract class ConfigSource {
+  ostype = Config.osType();
+  osarch = Config.osArch();
+  out_dir: string = Config.getSeleniumDir();
+  opt_ignoreSSL: boolean;
+  opt_proxy: string;
+
+  abstract getUrl(version: string): Promise<{url: string, version: string}>;
+  abstract getVersionList(): Promise<string[]>;
+}
+
+export abstract class XmlConfigSource extends ConfigSource {
+  xml: any;
+  xmlUrl: string;
+  fileName: string;
+
+  constructor(private name: string) {
+    super();
+  }
+
+  protected getFileName(): string {
+    try {
+      fs.statSync(this.out_dir);
+    } catch (e) {
+      fs.mkdirSync(this.out_dir);
+    }
+    this.fileName = path.resolve(this.out_dir, this.name + '-response.xml');
+    return this.fileName;
+  }
+
+  protected getXml(): Promise<any> {
+    this.getFileName();
+    let content = this.readResponse();
+    if (content != null) {
+      this.xml = content;
+      return Promise.resolve(this.xml);
+    }
+    return this.requestXml(this.xmlUrl, this.opt_ignoreSSL, this.opt_proxy).then(text => {
+      this.xml = this.convertXml2js(text);
+      fs.writeFileSync(this.fileName, text);
+      return this.xml;
+    });
+  }
+
+  private readResponse(): any {
+    try {
+      let contents = fs.readFileSync(this.fileName).toString();
+      let timestamp = new Date(fs.statSync(this.fileName).mtime).getTime();
+
+      let now = Date.now();
+      if (now - 36000000 < timestamp) {
+        return this.convertXml2js(contents);
+      }
+    } catch (err) {
+      return null;
+    }
+  }
+
+  private requestXml(url: string, opt_ignoreSSL: boolean, opt_proxy: string): Promise<string> {
+    return new Promise<string>((resolve, reject) => {
+      let options = HttpUtils.initOptions(url);
+      options = HttpUtils.optionsSSL(options, opt_ignoreSSL);
+      options = HttpUtils.optionsProxy(options, url, opt_proxy);
+
+      let req = request(options);
+      req.on('response', response => {
+        if (response.statusCode === 200) {
+          // logger.info('curl -v ' + options.url);
+          let output = '';
+          response.on('data', (data) => {
+            output += data;
+          });
+          response.on('end', () => {
+            resolve(output);
+          });
+
+        } else {
+          reject(new Error('response status code is not 200'));
+        }
+      });
+    });
+  }
+
+  private convertXml2js(xml: string): any {
+    let retResult: any = null;
+    xml2js.parseString(xml, (err, result) => {
+      retResult = result;
+    });
+    return retResult;
+  }
+}
+
+export abstract class JsonConfigSource extends ConfigSource {
+  json: any;
```

Comment:
As with the `XmlConfigSource` comments, the `json` and `fileName` properties should be removed, and the `jsonUrl` property needs a comment